core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1817,8 +1835,146 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+  // project should NOT be added to any Service Perimeters yet, that needs to happen AFTER we persist the Workspace
+  // record.
+  // 1. Claim Project from RBS
+  // 2. Update Billing Account information on Google Project
+  // Returns a Future[(GoogleProjectId, GoogleProjectNumber)] of the project that we claimed from RBS
+  private def setupGoogleProject(billingProject: RawlsBillingProject, span: Span = null): Future[(GoogleProjectId, GoogleProjectNumber)] = {
+    // We should never get here with a missing or invalid Billing Account, but we still need to get the value out of the
+    // Option, so we are being thorough
+    val billingAccount = billingProject.billingAccount match {
+      case Some(ba) if !billingProject.invalidBillingAccount => ba
+      case _ => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Billing Account is missing or invalid for Billing Project: ${billingProject}"))
+    }
+
+    // TODO: Wire up the Spans for tracing
+    for {
+      googleProjectId <- traceWithParent("getGoogleProjectFromRBS", span)(_ => getGoogleProjectFromRBS)
+      _ <- updateBillingAccountOnGoogleProject(googleProjectId, billingAccount)
+      googleProjectNumber <- getGoogleProjectNumber(googleProjectId)
+    } yield (googleProjectId, googleProjectNumber)
+  }
+
+  private def getGoogleProjectNumber(googleProjectId: GoogleProjectId): Future[GoogleProjectNumber] = {
+    gcsDAO.getGoogleProject(googleProjectId).map { p => Option(p.getProjectNumber) match {
+        case None => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, s"Failed to retrieve Google Project Number for Google Project ${googleProjectId}"))
+        case Some(longProjectNumber) => GoogleProjectNumber(longProjectNumber.toString)
+      }
+    }
+  }
+
+  // If a ServicePerimeter is specified on the BillingProject, then we will need to ensure that the provided Google
+  // Project is added to that Service Perimeter.  If no ServicePerimeter is specified on the Billing Project, do nothing
+  private def maybeAddGoogleProjectToPerimeter(billingProject: RawlsBillingProject, googleProjectNumber: GoogleProjectNumber, span: Span = null): Future[Unit] = {
+    billingProject.servicePerimeter match {
+      case Some(servicePerimeterName) => addGoogleProjectToPerimeter(googleProjectNumber, servicePerimeterName)
+      case None => Future.successful()
+    }
+  }
+
+  // Some Service Perimeters may required that they have some additional non-Terra Google Projects that need to be in
+  // the perimeter for some other reason.  They provide those to us and we add them to the Rawls Config so that
+  // whenever we update the list of projects for a perimeter, these projects are always included
+  private def loadStaticProjectsForPerimeter(servicePerimeterName: ServicePerimeterName): Seq[String] = {
+    val staticProjectsConfig = ConfigFactory.load().getConfig("gcs.servicePerimeters.staticProjects")
```

Comment:
config should be passed into the WorkspaceService via constructor args, not loaded on demand. This low form of dependency injection makes testing an reuse much easier.